pysph/cpy/parallel.py

```diff
@@ -107,6 +106,256 @@
+                % if use_segment:
+                if (scan_seg_flags[i]):
+                    a = ${neutral}
+                else:
+                    a = temp
+                % else:
+                a = temp
+                % endif
+
+                # Map
+                b = ${input_expr}
+                % if complex_map:
+                map_output[i] = b
+                % endif
+
+                # Scan
+                temp = ${scan_expr}
+
+            buffer[buffer_idx] = temp
+            % if use_segment:
+            chunk_new_segment[buffer_idx] = has_segment
+            % endif
+
+        # Pass 2: Aggregate chunks
+        # Add previous carry to buffer[0]
+        % if use_segment:
+        if chunk_new_segment[0]:
+            a = ${neutral}
+        else:
+            a = global_carry
+        b = buffer[0]
+        % else:
+        a = global_carry
+        b = buffer[0]
+        % endif
+        buffer[0] = ${scan_expr}
+
+        for i in range(n_thread - 1):
+            % if use_segment:
+
+            # With segmented scan
+            if chunk_new_segment[(i + 1) * stride]:
+                a = ${neutral}
+            else:
+                a = buffer[i * stride]
+            b = buffer[(i + 1) * stride]
+            buffer[(i + 1) * stride] = ${scan_expr}
+
+            % else:
+
+            # Without segmented scan
+            a = buffer[i * stride]
+            b = buffer[(i + 1) * stride]
+            buffer[(i + 1) * stride] = ${scan_expr}
+
+            % endif
+
+        last_item = buffer[(n_thread - 1) * stride]
+
+        # Shift buffer to right by 1 unit
+        for i in range(n_thread - 1, 0, -1):
+            buffer[i * stride] = buffer[(i - 1) * stride]
+
+        buffer[0] = global_carry
+        global_carry = last_item
+
+        # Pass 3: Output
+        with nogil, parallel():
+            tid = threadid()
+            buffer_idx = tid * stride
+            carry = buffer[buffer_idx]
+
+            start = offset + tid * chunksize
+            end = offset + min((tid + 1) * chunksize, SIZE)
+
+            for i in range(start, end):
+                # Output
+                % if use_segment:
+                if scan_seg_flags[i]:
+                    a = ${neutral}
+                else:
+                    a = carry
+                % else:
+                a = carry
+                % endif
+
+                % if complex_map:
+                b = map_output[i]
+                % else:
+                b = ${input_expr}
+                % endif
+
+                % if calc_prev_item:
+                prev_item = carry
+                % endif
+
+                carry = ${scan_expr}
+                item = carry
+
+                ${output_expr}
+        offset += chunksize * n_thread
+
+    # Clean up
+    free(buffer)
+
+    % if use_segment:
+    free(scan_seg_flags)
+    free(chunk_new_segment)
+    % endif
+
+    % if complex_map:
+    free(map_output)
+    % endif
+
```

Comment:
I wish it were possible to write this same thing with cpy, I think macros are important and need to think more about this.  Of course, `with` statements won't work and the nogil/parallel will need some work but the rest seems rather straightforward.  Its sad that we need n different implementations of this -- one for each backend.